package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -0,0 +1,690 @@
+
+        return True
+
+    cdef fill_grid(self, real[:, ::1] coords):
+        cdef ns_int i, cellindex = -1
+        cdef ns_int ncoords = coords.shape[0]
+        cdef ns_int[:] beadcounts = np.empty(self.size, dtype=np.int)
+
+        with nogil:
+            # Initialize buffers
+            for i in range(self.size):
+                beadcounts[i] = 0
+
+            # First loop: find cellindex for each bead
+            for i in range(ncoords):
+                cellindex = self.coord2cellid(&coords[i, 0])
+
+                self.nbeads[cellindex] += 1
+                self.cellids[i] = cellindex
+
+                if self.nbeads[cellindex] > self.nbeads_per_cell:
+                    self.nbeads_per_cell = self.nbeads[cellindex]
+
+        # Allocate memory
+        self.beadids = <ns_int *> PyMem_Malloc(sizeof(ns_int) * self.size * self.nbeads_per_cell) #np.empty((self.size, nbeads_max), dtype=np.int)
+        if not self.beadids:
+            raise MemoryError("Could not allocate memory for NSGrid.beadids ({} bits requested)".format(sizeof(ns_int) * self.size * self.nbeads_per_cell))
+
+        with nogil:
+            # Second loop: fill grid
+            for i in range(ncoords):
+
+                # Add bead to grid cell
+                cellindex = self.cellids[i]
+                self.beadids[cellindex * self.nbeads_per_cell + beadcounts[cellindex]] = i
+                beadcounts[cellindex] += 1
+
+
+
+cdef class FastNS(object):
+    cdef bint debug
+    cdef PBCBox box
+    cdef real[:, ::1] coords
+    cdef real[:, ::1] coords_bbox
+    cdef readonly real cutoff
+    cdef bint prepared
+    cdef NSGrid grid
+
+    def __init__(self, u, cutoff, coords=None, prepare=True, debug=False, max_gridsize=5000):
+        import MDAnalysis as mda
+        from MDAnalysis.lib.mdamath import triclinic_vectors
+
+        self.debug = debug
+
+        if not isinstance(u, mda.Universe):
+            raise TypeError("FastNS class must be initialized with a valid MDAnalysis.Universe instance")
+        box = triclinic_vectors(u.dimensions)
+
+        self.box = PBCBox(box)
+
+
+        if coords is None:
+            coords = u.atoms.positions
+
+        self.coords = coords.copy()
+
+        self.coords_bbox =  self.box.fast_put_atoms_in_bbox(coords)
+
+        if cutoff < 0:
+            raise ValueError("Cutoff must be positive!")
+        if cutoff * cutoff > self.box.c_pbcbox.max_cutoff2:
+            raise ValueError("Cutoff greater than maximum cutoff ({:.3f}) given the PBC")
+        self.cutoff = cutoff
+
+        self.grid = NSGrid(self.coords_bbox.shape[0], cutoff, self.box, max_gridsize, debug=debug)
+        self.prepared = False
+        if prepare:
+            self.prepare()
+
+
+    def prepare(self, force=False):
+        if self.prepared and not force:
+            return
+
+        self.grid.fill_grid(self.coords_bbox)
+
+        self.prepared = True
+
+    def search(self, search_ids=None):

```

Comment:
I would have preferred if it takes input coordinates i.e. ``u.atoms.positions`` as argument, but this also works (specially considering its use with universe instances.